### PR TITLE
Converted ZC_CHANGE_CHATROOM to struct

### PR DESCRIPTION
**Pull request by AoShinRO**
* **Addressed Issue(s)**: 

* **Server Mode**: both

* **Description of Pull Request**: Converted ZC_CHANGE_CHATROOM to struct

### CODE REVIEW

File: src/map/clif.cpp [hunk 1]
```diff
@@ -4486,28 +4486,43 @@ void clif_dispchat(struct chat_data* cd, int fd)
...
+	p->packetSize = static_cast<decltype(p->packetSize)>(17 + strlen(cd.title));
+	p->ownerId = cd.owner->id;
+	p->chatId = cd.bl.id;
+	p->Limit = cd.limit;
+	p->Users = cd.users;
+
+	memcpy(p->Title, cd.title, strlen(cd.title)); // not zero-terminated
+
+	uint8 type = CHAT_PUBLIC;
+
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:4513]:
```suggestion
```

File: src/map/clif.cpp [hunk 2]
```diff
@@ -4486,28 +4486,43 @@ void clif_dispchat(struct chat_data* cd, int fd)
...
+	if(cd.owner->type == BL_NPC){
+		p->Users++;
+		if(cd.limit)
+			type = CHAT_ARENA;
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:4517]:
```suggestion
			p->flag = CHAT_ARENA;
```

File: src/map/clif.cpp [hunk 3]
```diff
@@ -4486,28 +4486,43 @@ void clif_dispchat(struct chat_data* cd, int fd)
...
+		else
+			type = CHAT_PK;
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:4519]:
```suggestion
			p->flag = CHAT_PK;
```

File: src/map/clif.cpp [hunk 4]
```diff
@@ -4486,28 +4486,43 @@ void clif_dispchat(struct chat_data* cd, int fd)
...
+	}else if(cd.owner->type == BL_PC && cd.pub == false)
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:4520]:
```suggestion
	}else if(cd.owner->type == BL_PC && cd.pub == false){
```

File: src/map/clif.cpp [hunk 5]
```diff
@@ -4486,28 +4486,43 @@ void clif_dispchat(struct chat_data* cd, int fd)
...
+		type = CHAT_PRIVATE;
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:4521]:
```suggestion
		p->flag = CHAT_PRIVATE;
	}else{
		p->flag = CHAT_PUBLIC;
	}
```

File: src/map/clif.cpp [hunk 6]
```diff
@@ -4486,28 +4486,43 @@ void clif_dispchat(struct chat_data* cd, int fd)
...
+
+	p->Flag = type;
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:4523]:
```suggestion
```

File: src/map/clif.hpp [hunk 7]
```diff
@@ -889,7 +889,7 @@ void clif_addchat(struct chat_data* cd,map_session_data *sd);	// chat
 void clif_changechatowner(struct chat_data* cd, map_session_data* sd);	// chat
 void clif_clearchat(struct chat_data *cd,int fd);	// area or fd
 void clif_leavechat(struct chat_data* cd, map_session_data* sd, bool flag);	// chat
-void clif_changechatstatus(struct chat_data* cd);	// chat
+void clif_changechatstatus(chat_data& cd);	// chat
```
Comment (Lemongrass3110) [inline src/map/clif.hpp:892]:
```suggestion
void clif_changechatstatus(chat_data& cd);
```

File: src/map/packets.hpp [hunk 8]
```diff
@@ -838,6 +838,18 @@ struct PACKET_ZC_PAR_CHANGE_USER {
...
 
+struct PACKET_ZC_CHANGE_CHATROOM{
+	uint16 packetType;
+	uint16 packetSize;
+	uint32 ownerId;
+	uint32 chatId;
+	uint16 Limit;
+	uint16 Users;
+	uint8 Flag;
+	char Title[];
```
Comment (Lemongrass3110) [inline src/map/packets.hpp:849]:
```suggestion
	uint16 limit;
	uint16 users;
	uint8 flag;
	char title[];
```

File: src/map/clif.cpp [hunk 9]
```diff
@@ -4486,28 +4486,43 @@ void clif_dispchat(struct chat_data* cd, int fd)
...
+	PACKET_ZC_CHANGE_CHATROOM* p = reinterpret_cast<PACKET_ZC_CHANGE_CHATROOM*>( packet_buffer );
+
+	p->packetType = HEADER_ZC_CHANGE_CHATROOM;
+	p->packetSize = static_cast<decltype(p->packetSize)>(17 + strlen(cd.title));
+	p->ownerId = cd.owner->id;
+	p->chatId = cd.bl.id;
+	p->Limit = cd.limit;
+	p->Users = cd.users;
+
+	memcpy(p->Title, cd.title, strlen(cd.title)); // not zero-terminated
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:4510]:
```suggestion
	// not zero-terminated
	strncpy(p->title, cd.title, strlen(cd.title));
```

File: src/map/clif.cpp [hunk 10]
```diff
@@ -4486,28 +4486,43 @@ void clif_dispchat(struct chat_data* cd, int fd)
...
-	WBUFB(buf,16) = type;
-	memcpy(WBUFCP(buf,17), cd->title, strlen(cd->title)); // not zero-terminated
+	PACKET_ZC_CHANGE_CHATROOM* p = reinterpret_cast<PACKET_ZC_CHANGE_CHATROOM*>( packet_buffer );
+
+	p->packetType = HEADER_ZC_CHANGE_CHATROOM;
+	p->packetSize = static_cast<decltype(p->packetSize)>(17 + strlen(cd.title));
+	p->ownerId = cd.owner->id;
+	p->chatId = cd.bl.id;
+	p->Limit = cd.limit;
+	p->Users = cd.users;
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:4508]:
```suggestion
	p->limit = cd.limit;
	p->users = cd.users;
```

File: src/map/clif.cpp [hunk 11]
```diff
@@ -4486,28 +4486,43 @@ void clif_dispchat(struct chat_data* cd, int fd)
...
+
+	memcpy(p->Title, cd.title, strlen(cd.title)); // not zero-terminated
+
+	uint8 type = CHAT_PUBLIC;
+
+	if(cd.owner->type == BL_NPC){
+		p->Users++;
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:4515]:
```suggestion
		// NPC itself counts as additional chat user
		p->users++;

```

**Review by Lemongrass3110 — APPROVED** [review]

File: src/map/clif.cpp [hunk 12]
```diff
@@ -4486,28 +4486,45 @@ void clif_dispchat(struct chat_data* cd, int fd)
...
+void clif_changechatstatus(chat_data& cd) {
 
-	if( cd == nullptr || cd->usersd[0] == nullptr )
+	if(cd.usersd[0] == nullptr )
 		return;
 
-	type = (cd->owner->type == BL_PC ) ? (cd->pub) ? 1 : 0
-	     : (cd->owner->type == BL_NPC) ? (cd->limit) ? 2 : 3
-	     : 1;
+	enum chat_flags:uint8 {
```
Comment (aleos89) [inline src/map/clif.cpp:4494]:
```suggestion
	enum e_chat_flags:uint8 {
```
I know it's scoped to this function, but just to keep up with the rAthena standard for naming!

**Review by aleos89 — APPROVED** [review]

**Pull request merged**